actors/builtin/market/market_actor.go

```diff
@@ -77,7 +78,11 @@ func (a Actor) WithdrawBalance(rt Runtime, params *WithdrawBalanceParams) *adt.E
 		rt.Abortf(exitcode.ErrIllegalArgument, "negative amount %v", params.Amount)
 	}
 
-	nominal, recipient := escrowAddress(rt, params.ProviderOrClientAddress)
+	nominal, recipient, codeId := escrowAddress(rt, params.ProviderOrClientAddress)
+	// if caller is not a miner, caller should be the recipient of the funds
+	if !codeId.Equals(builtin.StorageMinerActorCodeID) {
```

Comment:
Does the miner actor need to withdraw? I don't see any calls from the miner code.

Reply:
The escrowAddress already invokes `ValidateImmediateCallerXxx`. It's wrong to use that again here, though sometimes there is a good reason to manually check some other predicate.

The bluntest fix here is simply to change `ValidateImmediateCallerType` on line 701 to be `ValidateImmediateCallerIs` the recipient. That resolves the issue correctly, but would have the possibly-unintended consequence of also limiting AddBalance.

If we wanted to distinguish those, we would need to pull the last half of `escrowAddress` up to the callers. Perhaps it could return a tuple of `(nominal Address, approvedCallers []Address)`, and we pull the caller validation up (ignoring approved callers for AddBalance)

Reply:
This is done.